File: src/main/scala/alankstewart/math/lineart/LineComponent.scala
package alankstewart.math.lineart

import java.awt._
import java.awt.geom.Line2D
import javax.swing.JComponent

import scala.swing.Graphics2D

/**
  * Created by alanstewart on 22/02/2016.
  */
abstract class LineComponent(d: Dimension) extends JComponent {

  val tx = d.getWidth / 2.0
  val ty = d.getHeight / 2.0

  override def paintComponent(g: Graphics) = {
    val g2d = g match {
      case g2d: Graphics2D => g2d
      case _ => throw new ClassCastException
    }
    g2d.setColor(Color.DARK_GRAY)
    g2d.translate(tx, ty)
    g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON)
    Range.inclusive(1, getLines).map(getLineSegment).foreach(s => g2d.draw(s))
  }

  def getLines: Int

  def getLineSegment(n: Int): Line2D.Double
}

File: src/main/java/alankstewart/math/lines/LineComponent.java
package alankstewart.math.lines;

import javax.swing.*;
import java.awt.*;
import java.awt.geom.Line2D;
import java.util.stream.IntStream;

/**
 * Created by alanstewart on 23/01/15.
 */
abstract class LineComponent extends JComponent {

    protected final double tx;
    protected final double ty;

    public LineComponent(final Dimension d) {
        tx = d.getWidth() / 2.0;
        ty = d.getHeight() / 2.0;
    }

    @Override
    public void paintComponent(final Graphics g) {
        final Graphics2D g2d = (Graphics2D) g;
        g2d.setColor(Color.DARK_GRAY);
        g2d.translate(tx, ty);
        g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
        IntStream.rangeClosed(1, getLines()).mapToObj(this::getLineSegment).forEach(g2d::draw);
    }

    protected abstract int getLines();

    protected abstract Line2D.Double getLineSegment(final int n);
}
